### PR TITLE
While the Script Editor Plugin is active, makes it optional to switch Editor Plugins when selecting Nodes in the Scene Tree

### PR DESCRIPTION
**Pull request by alfredbaudisch**
## Problem
- **While the Script editor is active, selecting a Node in the Scene Tree in order to inspect it, makes the `editor_plugin_screen` to be changed to the Node's main plugin**.
- The same happens when right-clicking a Node
- This makes it hard, and tiresome to **inspect and contextualize Nodes while writing Scripts**, because it **forcibly causes unintentional repetitive viewport and editor switching when interacting with Nodes while in the Script Editor**.
- It makes it even more cumbersome when dealing with long UI-centric Scene Trees, where it's necessary to gather Node Paths and adjust Inspector properties
- **The current behaviour is more than an annoyance than a desired effect** - I never faced any use case where I wanted to switch viewports when clicking a Node while in the Script Editor - **you end up FORCIBLY SWITCHING CONTEXTS back hundreds of times in a single day of work**. If I want to switch to the 2D or 3D viewports I can clearly click their toolbar buttons or use the switch hotkeys.
- It also makes the process of dragging and dropping Nodes into the Text Editor repetitive and tiresome, requiring 4 clicks: (1) click a Node, it switches to the 2D/3D viewport, (2) click to go back to Script, (3) click the Node again, (4) drag and drop it into the Text Editor to get the Node Path.
- Examples are reported in this [blog post](https://alfredbaudisch.com/blog/gamedev/godot-engine/godots-3d-confusing-workflow-inconsistencies-conflicting-behaviours-and-annoyances/), under "Inconsistent Viewport and Script Editor tab switching when clicking Nodes (2D/3D)".

https://user-images.githubusercontent.com/248383/180605492-0e974253-ec46-474d-b3cf-f9718fb3fa2f.mp4

## Solution
- **Make it optional to switch Editor Plugins when selecting Nodes** in the Scene tree **while the current active Editor Plugin is the Script Text Editor**. 
- It **works both with the Left Click** (PRIMARY EFFECT: to access the Node's Inspector) **and with Right Click** (SECONDARY EFFECT: to access the Node's Context Menu), both more than common and repetitive behaviors needed while writing Scripts.
- The feature is toggeable via `text_editor/behavior/navigation/node_select_toggles_inspector_only`.

https://user-images.githubusercontent.com/248383/180605429-7fa86418-e554-4386-afd7-fad42e3763d3.mp4

## Is it really needed?
- I make complex user interfaces with Godot, such as my project [Godello](https://github.com/alfredbaudisch/Godello) and [Dynamic Inventory System](https://github.com/alfredbaudisch/GodotDynamicInventorySystem), and this has always been in my way when scripting the complex UI Node trees.
- My tweet on this subject had a [few answers and retweets](https://twitter.com/AlfredBaudisch/status/1550770153066995712) regarding this to be a welcome addition, including [an answer](https://twitter.com/MagellanicGames/status/1550788937211416576) from the prolific Godot's developer MagellanicGames. Also [here](https://twitter.com/Saulcava1/status/1550837632179077126) and [here](https://twitter.com/pigdev/status/1550827749861908485).
- Update: many more public responses: [here](https://twitter.com/Jyve/status/1550869231838822400), [here](https://twitter.com/GreenMoonmoon/status/1550885913022054404), [here](https://twitter.com/fagnerln/status/1550914068529152001), [here](https://twitter.com/Seglectic/status/1550980186316865536), [here](https://twitter.com/coelhucass/status/1551021160598568960) (and other RTs, etc).

### CODE REVIEW

**Comment by triptych** [comment]
Please consider adding this!

**Comment by minuJeong** [comment]
i want this!

**Comment by mhilbrunner** [comment]
Please refrain from adding comments that don't add to the discussion of this change. If you just want to express your support for this change, use the reaction on the PR itself for that. Thank you!

(Comments add some noise for maintainers who then see the PR has received an update; some also get an email about it. It also makes discussion harder to follow in longer PR comment chains.)

**Comment by samdze** [comment]
I don't know if I like this solution, feels not optimal to me.
I'd better like a unified solution that works best almost all the time since the default behaviour is still preferred occasionally.

What about the switching only happening if an already selected node gets clicked again? (click down + release, to allow dragging it without a switch)

**Comment by Calinou** [comment]
Note that this is already available when holding <kbd>Alt</kbd> while clicking in the scene tree dock. Unfortunately, most window managers on Linux will intercept <kbd>Alt</kbd> + clicking by default, so Godot will never see those clicks.

**Comment by KoBeWi** [comment]
The Alt was a hack that got removed in #61162

I think the proper solution would be extending #61162 to handle script editor properly (i.e. only switch editor when not dragging the node).

**Comment by alfredbaudisch** [comment]
> (i.e. only switch editor when not dragging the node).
If it requires you to drag the node to not switch the editor, how would you inspect the Nodes in the Inspector and access their context menu, while staying in the Script editor?

Wouldn't that require you to loosely always drag Nodes and let them go, without the intention of actually dragging, but just because you want to access the Inspector for that Node?

The two biggest pain points to be fixed are getting distraction-free and hassle-free **access to the Inspector and access to the context menu of any Node, while staying in the Script Editor, without toggling a context switching**.

**Comment by KoBeWi** [comment]
Ok it makes sense to prevent switching on node click.

> Isn't the default behaviour more of a legacy leftover than actually being useful?

idk I sometimes use it, but I agree that more often it's annoying and *not switching* is more useful.

> What about the switching only happening if an already selected node gets clicked again?

Or maybe we could switch when Alt is held or something.

But editor setting is fine too probably; it should be enabled by default though. Also the name could be better, like `stay_in_script_editor_on_node_selected` (up to discussion).

File: editor/editor_node.cpp
Comment (KoBeWi) [inline editor/editor_node.cpp:2218]:
```C++
inspector_only = !inspector_only
```
If `inspector_only` is true, this will always disable it.

```C++
editor_plugin_screen != nullptr && editor_plugin_screen->get_name() == "Script"
```
Instead of this, you could do `ScriptEditor::get_singleton()->is_visible_in_tree()` ([that's how it was done previously](https://github.com/godotengine/godot/blob/838cb598e3ef5572018d3d549c15bb27b0a5864d/editor/scene_tree_dock.cpp#L1384)).

**Comment by 8to16apps** [comment]
Hi all, first comment event in github.

I would like introduce another posible solution.
A pin icon in the script tool bar, after File menu o next to rigth buttons,
so you can pin to make script visible even if change the node or unpin to have the current behavior.
This make this function more visible and move easy to switch on/off

File: editor/editor_node.cpp
Comment (alfredbaudisch) [reply editor/editor_node.cpp:2218]: `inspector_only = !inspector_only`

My intention here is to not override `inspector_only` when it's already `true`.

Say `node_select_toggles_inspector_only == false` and `inspector_only == true`, then `inspector_only` will be overriden to `false` (this is based on the assumption that there are already situations where `inspector_only` may be true).

With my intention, we try to override it only when a context switching was already going to take place (`inspector_only == false`).

---

`ScriptEditor::get_singleton()->is_visible_in_tree()`

Cool, thanks.

Comment (alfredbaudisch) [reply editor/editor_node.cpp:2218]: > stay_in_script_editor_on_node_selected 

@KoBeWi I also prefer your suggestion. Also, do you think `text_editor/behavior/navigation` is still the best category to put it?

Comment (KoBeWi) [reply editor/editor_node.cpp:2218]: > My intention here is to not override inspector_only when it's already true.

Except with the current code, `inspector_only` will always be turned false if other conditions fail, because you used only `&&`.

I think the condition should be something like
```
inspector_only = inspector_only || (node_select_toggles_inspector_only && ...);
```

> do you think text_editor/behavior/navigation is still the best category to put it?

From the existing categories in `text_editor` group this one is the best.

Comment (alfredbaudisch) [reply editor/editor_node.cpp:2218]: Sorry, I thought it was already there, I fixed exactly that (`||`) yesterday and haven't pushed it.

> From the existing categories in text_editor group this one is the best.
Cool, I'll push with the new name and the logic fix.

**Comment by reduz** [comment]
I agree with@8to16apps and also think a pin icon in the script toolbar is definitely a cleaner solution. Very discoverable and easy to toggle on and off.

**Comment by alfredbaudisch** [comment]
@reduz @8to16apps I'm not sure I understand where exactly it should be. A mockup would be welcome.

**Comment by 8to16apps** [comment]
@alfredbaudisch @reduz I am thinking in something like this. This is the same concept that the animation tap in the bottom.

mock_up

**Comment by alfredbaudisch** [comment]
@8to16apps Good point. Doing this way would keep it consistent with the pin behaviour of the AnimationPlayer.


image

**Comment by KoBeWi** [comment]
btw the pin button can simply toggle the editor setting, so you can keep your current changes.

**Comment by alfredbaudisch** [comment]
> btw the pin button can simply toggle the editor setting, so you can keep your current changes.

You're fast as I was about to comment: my implementation wouldn't change.
The Pin would simply be a shortcut to toggle my new setting.

**Comment by reduz** [comment]
We put pins on the right side, so I guess I would put it on the right side here too.

**Comment by alfredbaudisch** [comment]
@reduz @8to16apps Done

https://user-images.githubusercontent.com/248383/180646125-b20d86cd-196b-4c9c-8bfa-2012599b5b5b.mp4

File: editor/plugins/script_editor_plugin.cpp
Comment (KoBeWi) [inline editor/plugins/script_editor_plugin.cpp:3431]: You should save the settings, otherwise this won't be remembered.

Comment (alfredbaudisch) [reply editor/plugins/script_editor_plugin.cpp:3431]: Thanks. I assumed it was already permanent.

File: editor/editor_node.cpp
Comment (YeldhamDev) [inline editor/editor_node.cpp:2218]:
```suggestion
			if (!inspector_only) {
```

**Review by YeldhamDev — CHANGES_REQUESTED** [review]

**Comment by YeldhamDev** [comment]
The commits need to be squashed before the PR can be merged. You can follow the instructions for that here: https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase

You can use `git commit --amend` in the subsequent commits to avoid this process.

**Comment by alfredbaudisch** [comment]
Done

**Review by YeldhamDev — APPROVED** [review]

**Comment by YuriSizov** [comment]
The pin feels like a lot of micromanagement for something that should probably work one way out of the box. Like Emi mentioned on Twitter, I'd probably get annoyed by the pin button wasting space and remove it with a plugin. I don't believe it's something you need to constantly change to justify having it so visible.

**Comment by alfredbaudisch** [comment]
> The pin feels like a lot of micromanagement for something that should probably work one way out of the box. Like Emi mentioned on Twitter, I'd probably get annoyed by the pin button wasting space and remove it with a plugin. I don't believe it's something you need to constantly change to justify having it so visible.

I'm neutral about it actually, since my intended solution is solved with or without the pin. So how things like this are normally decided? @YuriSizov

**Comment by YuriSizov** [comment]
> So how things like this are normally decided? 

I guess we'll discuss it during the next PR review meeting. You are welcome to join, it's on Tuesday (watch out in #devel channel on RocketChat).

**Comment by alfredbaudisch** [comment]
The pin has been removed in favor of keeping the first basic implementation of just the Editor setting (on by default), as per some [Twitter conversation](https://twitter.com/reduzio/status/1551554651169996800).

**Review by akien-mga — APPROVED** [review]
Looks good to me.

Not 100% convinced by the setting name but this could be improved later on.

**Comment by akien-mga** [comment]
Thanks!

**Comment by Anutrix** [comment]
Can this be cherry-picked to 3.x branch?

**Comment by akien-mga** [comment]
Cherry-picked for 3.5.

@alfredbaudisch Just noticed it but for future commits, please try to keep the commit message title (first line) at a reasonable length. It should typically be less than 72 chars, the current one is 255 (i.e. your commit title is a full description instead of a title).

**Comment by alfredbaudisch** [comment]
@akien-mga Thanks Rémi for fixing the commit message and for the tips! I'll keep those conventions in mind for possible future contributions.

**Pull request merged**
